### PR TITLE
[FDB] Fix fbdorch to properly handle syncd FDB FLUSH Notif 

### PR DESCRIPTION
**Pull request by vivekreddynv**
**What I did**

Made enhancements to overcome all the below mentioned shortcomings

**Why I did it**

1) Vlan delete couldn't be handled by OA when there is fdb learnt on the member and when the member is deleted
2) This inability of handling APPL_DB notif is affecting warm-restart
3) FDB Entry from State DB is not removed.
4) OA doesn't have the logic to handle consolidate flush notif coming from syncd
5) FdbOrch doesn't have logic to clear internal cache and decrement corresponding fdb counters during a flush notification

**How I verified it**

Checked this config and verified the remove vlan passed though
```
1. config vlan add 40
2. config vlan member add 40 Ethernet0
On the other side. 
3. sudo ip link add link enp4s0f0 name enp4s0f0.40 type vlan id 40
4. sudo ip link set enp4s0f0.40 up
5. Make sure a fdb entry is learnt. 

root@sonic:/home/admin# fdbshow
  No.    Vlan  MacAddress         Port       Type
-----  ------  -----------------  ---------  -------
    1      40  7C:FE:90:12:22:EC  Ethernet0  Dynamic

dump state fdb all --db STATE_DB
{
    "Vlan40:7c:fe:90:12:22:ec": {
        "STATE_DB": {
            "keys": [
                {
                    "FDB_TABLE|Vlan40:7c:fe:90:12:22:ec": {
                        "port": "Ethernet0",
                        "type": "dynamic"
                    }
                }
            ],
            "tables_not_found": []
        }
    }
}

7. config vlan member del 40 Ethernet0
8. config vlan del 40
9. dump state fdb all --db STATE_DB
{}
```
**Unit Tests:**
```
vkarri@4897dcba242c:/sonic/src/sonic-swss/tests/mock_tests$ ./tests --gtest_filter=FdbOrchTest*
Running main() from /usr/src/gtest/src/gtest_main.cc
Note: Google Test filter = FdbOrchTest*
[==========] Running 6 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 6 tests from FdbOrchTest
[ RUN      ] FdbOrchTest.ConsolidatedFlushVlanandPort
[       OK ] FdbOrchTest.ConsolidatedFlushVlanandPort (14 ms)
[ RUN      ] FdbOrchTest.ConsolidatedFlushAll
[       OK ] FdbOrchTest.ConsolidatedFlushAll (15 ms)
[ RUN      ] FdbOrchTest.ConsolidatedFlushVlan
[       OK ] FdbOrchTest.ConsolidatedFlushVlan (14 ms)
[ RUN      ] FdbOrchTest.ConsolidatedFlushPort
[       OK ] FdbOrchTest.ConsolidatedFlushPort (15 ms)
[ RUN      ] FdbOrchTest.ConsolidatedFlushVlanandPortBridgeportDeleted
[       OK ] FdbOrchTest.ConsolidatedFlushVlanandPortBridgeportDeleted (14 ms)
[ RUN      ] FdbOrchTest.NonConsolidatedFlushVlanandPort
[       OK ] FdbOrchTest.NonConsolidatedFlushVlanandPort (15 ms)
[----------] 6 tests from FdbOrchTest (87 ms total)

[----------] Global test environment tear-down
[==========] 6 tests from 1 test case ran. (87 ms total)
[  PASSED  ] 6 tests.

```

**Details if related**

### CODE REVIEW

**Review by dgsudharsan — APPROVED** [review]

**Comment by vivekreddynv** [comment]
@dgsudharsan please review

File: orchagent/fdborch.cpp [hunk 1]
```diff
@@ -174,41 +173,155 @@ bool FdbOrch::storeFdbEntryState(const FdbUpdate& update)
...
+        Port temp_vlan;
+        if (m_portsOrch->getPort(update.entry.bv_id, temp_vlan))
+        {
+            temp_vlan.m_fdb_count--;
+            m_portsOrch->setPort(temp_vlan.m_alias, temp_vlan);
+        }
+
+        /* Decrement port fdb_counter */
+        Port temp_port;
+        if (m_portsOrch->getPort(update.port.m_alias, temp_port))
```
Comment (dgsudharsan) [inline orchagent/fdborch.cpp:273]: Why do you need to get port here again when you already obtained it previously. You just need to skip the zero port alone here

File: tests/mock_tests/fdborch/flush_syncd_notif_ut.cpp [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+                { APP_VLAN_TABLE_NAME, portsorch_base_pri + 2 },
+                { APP_VLAN_MEMBER_TABLE_NAME, portsorch_base_pri },
+                { APP_LAG_TABLE_NAME, portsorch_base_pri + 4 },
+                { APP_LAG_MEMBER_TABLE_NAME, portsorch_base_pri }
+            };
+
+            m_portsOrch = std::make_shared<PortsOrch>(m_app_db.get(), m_state_db.get(), ports_tables, m_chassis_app_db.get());
+
+            // 2) Crmorch
+            ASSERT_EQ(gCrmOrch, nullptr);
```
Comment (dgsudharsan) [inline tests/mock_tests/fdborch/flush_syncd_notif_ut.cpp:71]: Why do we need an assert here?

File: tests/mock_tests/fdborch/flush_syncd_notif_ut.cpp [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+                      m_portsOrch->m_portList[VLAN40].m_vlan_info.vlan_oid, SAI_FDB_ENTRY_TYPE_DYNAMIC);
+
+        /* make sure fdb_counter for Vlan is decremented */
+        ASSERT_EQ(m_portsOrch->m_portList[VLAN40].m_fdb_count, 0);
+        ASSERT_EQ(m_portsOrch->m_portList[ETH0].m_fdb_count, 0);
+
+        /* Make sure state db is cleared */
+        ASSERT_EQ(m_fdborch->m_fdbStateTable.hget("Vlan40:7c:fe:90:12:22:ec", "port", port), false);
+        ASSERT_EQ(m_fdborch->m_fdbStateTable.hget("Vlan40:7c:fe:90:12:22:ec", "type", entry_type), false);
+    }
```
Comment (dgsudharsan) [inline tests/mock_tests/fdborch/flush_syncd_notif_ut.cpp:380]: Can you add one test case for non consolidated mac, which has vlan and port?

**Review by dgsudharsan — CHANGES_REQUESTED** [review]

File: orchagent/fdborch.cpp [hunk 4]
```diff
@@ -174,41 +173,155 @@ bool FdbOrch::storeFdbEntryState(const FdbUpdate& update)
...
+*/
+void FdbOrch::handleSyncdFlushNotif(const sai_object_id_t& bv_id,
+                                    const sai_object_id_t& bridge_port_id,
+                                    const string& fdb_type,
+                                    const MacAddress& mac)
+{
+    MacAddress flush_mac(CONSOLIDATED_FLUSH_MAC);
+
+    /* Populate Queue */
+    vector<FdbUpdate> updates;
```
Comment (dgsudharsan) [inline orchagent/fdborch.cpp:190]: Please avoid storing it in vector. The Port is a big Class and if there are 10K entries on a port for which flush it called this vector will spike the memory.
Instead call the storeFdbEntryState ,notify and ref count decrement while you iterate through the entries.

**Review by dgsudharsan — CHANGES_REQUESTED** [review]

File: tests/mock_tests/fdborch/flush_syncd_notif_ut.cpp (hunk 2 above)
Comment (vivekreddynv) [reply tests/mock_tests/fdborch/flush_syncd_notif_ut.cpp:71]: Not very important, but kinda notifies if any previous tests did not clear/free this

File: tests/mock_tests/fdborch/flush_syncd_notif_ut.cpp (hunk 3 above)
Comment (vivekreddynv) [reply tests/mock_tests/fdborch/flush_syncd_notif_ut.cpp:380]: Added

**Review by dgsudharsan — APPROVED** [review]

**Pull request closed without merging**